File: leetcode-04/solutions/cpp/0748.cpp
class Solution {
 public:
  string shortestCompletingWord(string licensePlate, vector<string>& words) {
    string ans;

    vector<int> map(26);
    for (char c : licensePlate)
      if (isalpha(c)) map[tolower(c) - 'a']++;

    int min = INT_MAX;

    for (string& word : words) {
      if (word.length() >= min) continue;
      if (!isMatch(word, map)) continue;
      min = word.length();
      ans = word;
    }

    return ans;
  }

 private:
  bool isMatch(string& word, vector<int>& map) {
    vector<int> wordMap(26);
    for (char c : word) wordMap[c - 'a']++;

    for (int i = 0; i < 26; i++)
      if (wordMap[i] < map[i]) return false;

    return true;
  }
};

File: leetcode-04/solutions/python3/0748.py
class Solution:
    def shortestCompletingWord(self, licensePlate: str, words: List[str]) -> str:
        def isMatch(word: str) -> bool:
            wordDict = collections.Counter(word)

            for i in string.ascii_lowercase:
                if wordDict[i] < dict[i]:
                    return False

            return True

        dict = {chr(i + ord('a')): 0 for i in range(26)}
        for c in licensePlate:
            if c.isalpha():
                dict[c.lower()] += 1

        min = float('inf')

        for word in words:
            if len(word) >= min:
                continue
            if not isMatch(word):
                continue
            min = len(word)
            ans = word

        return ans
